pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	var deletedRV string
+	if options.Preconditions != nil && options.Preconditions.ResourceVersion != nil {
+		deletedRV = *options.Preconditions.ResourceVersion
+	}
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, err
+	}
+
+	deletedTyped := deletedObj.(*oauthapi.UserOAuthAccessToken)
+	if !isValidUserToken((*oauthapi.OAuthAccessToken)(deletedTyped), ctxUserName) {
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	// don't mutate the original options
+	newOpts := options.DeepCopy()
+
+	switch {
+	case newOpts.Preconditions == nil:
+		newOpts.Preconditions = &metav1.Preconditions{UID: &deletedTyped.UID}
+
+	case newOpts.Preconditions != nil && newOpts.Preconditions.UID != nil && *newOpts.Preconditions.UID != deletedTyped.UID:
+		// the UID seeked to be deleted no longer exists/never existed
+		return &metav1.Status{Status: metav1.StatusFailure}, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+
+	default:
+		// Make sure the UID of the just-checked object matches the object we're going to remove
+		newOpts.Preconditions.UID = &deletedTyped.UID
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, newOpts)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		klog.V(4).Infof("no user found in the context")
+		return "", kerrors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
+	}
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		klog.V(4).Info("user in context has zero-length username")
+		return "", kerrors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
+	}
+
+	return userName, nil
+}
+
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList, username string) *oauthapi.UserOAuthAccessTokenList {
+	ret := oauthapi.UserOAuthAccessTokenList{}
+	for _, t := range l.Items {
+		if !isValidUserToken(&t, username) {
+			continue
+		}
+		ret.Items = append(ret.Items, oauthapi.UserOAuthAccessToken(t))
+	}
+
+	// when watching a resource, the client usually attempts to list it first and
+	// it then continues to watch from the last known RV
+	// This prevents the initial objects to be returned twice to the client
+	// (objects from the list + from the next watch with an uninitialized from-RV)
+	ret.ResourceVersion = l.ResourceVersion
+	return &ret
+}
+
+func userFieldSelectorToListOptions(opts *metainternal.ListOptions, userName string) (*metainternal.ListOptions, error) {
+	var newOpts *metainternal.ListOptions
+	if opts == nil {
+		newOpts = &metainternal.ListOptions{}
+	} else {
+		newOpts = opts.DeepCopy()
+	}
+
+	newOpts.FieldSelector = fields.AndSelectors(
+		fields.Set{"userName": userName}.AsSelector(),
```

Comment:
we need an e2e test for this

Reply:
I added some e2e tests here: https://github.com/openshift/origin/pull/25697

Reply:
moved e2e tests to this repo